### DISCUSSION TITLE
Make `error` codes non-optional for better error handling on the client side

### DISCUSSION THREAD

**Discussion by VValiente** [supabase/supabase#22671] [Feature Requests]
Upon implementing the authentication, I encountered `AuthError` and `APIError` types with properties having inconsistent availability. Sometimes, only `code` is available, other times, only `msg`. In this example, `error` and `errorDescription` are present but not the `code` and `msg`. 
```
Auth.AuthError.APIError(msg: nil, code: nil, error: Optional("invalid_grant"), errorDescription: Optional("Invalid login credentials"), weakPassword: nil))
```
This is why many frontend developers resolve to generic error messages because backend error message handling is often a mess.

Trying to identify different types of errors is such a pain and I've only implemented the Auth part. I already have a long list of conditions trying to catch which property has value. And even though there is sometimes human-readable error description, I assume they are not localized, are only intended for us developers, and are not meant to be shown directly to users.

I also cannot find a list of possible errors on the docs page. Or did I just miss it? If I did, can someone please tag me with the links? thank you. 

**PROPOSAL**
Make `error` non-optional, or whatever property that can help identify an error consistently, with `errorDescription` to help us understand what the error is about. This will help us map backend errors to a localized message in our app.

**Comment by adao** [comment]
I am also curious abot error handling, and I have not been able to find a reference for the REST API regarding what errors get returned. I just started using supabase and I keep finding myself writing `const { data, error } = await supabase ...` but not really doing anything with the error, besides `if (error) throw error`, because I don't know what its return values are.
